### PR TITLE
Context

### PR DESCRIPTION
**Pull request by paulyea**
allow contexts to be passed around in gorm.

- add new functions <function name>Context that's the same as <function name> except it takes in context.
- change the original function to call the context version of the function with context.background

### CODE REVIEW

**Review by codeferret1 — APPROVED** [review]
This seems legit, but how do we get a patch to try to push upstream given we added godep and changed a bunch of places in the code to point at the od version of this lib?

File: main.go [hunk 1]
```diff
@@ -621,10 +630,10 @@ func (s *DB) AddUniqueIndex(indexName string, columns ...string) *DB {
 	return scope.db
 }
 
-// RemoveIndex remove index with name
-func (s *DB) RemoveIndex(indexName string) *DB {
+// RemoveIndexContext remove index with name
+func (s *DB) RemoveIndexContext(indexName string) *DB {
 	scope := s.NewScope(s.Value)
-	scope.removeIndex(indexName)
+	scope.RemoveIndexContext(indexName)
```
Comment (ts4z) [inline main.go:636]: why did this go from calling a private method to a public one?

File: main_test.go [hunk 2]
```diff
@@ -195,18 +195,18 @@ func (c Cart) TableName() string {
 	return "shopping_cart"
 }
 
-func TestHasTable(t *testing.T) {
+func TestHasTableContext(t *testing.T) {
```
Comment (ts4z) [inline main_test.go:198]: this is going to be a hard sell to merge upstream. The change to the test is OK, but now this guts the test for the old flavor of APIs.

File: scope.go [hunk 3]
```diff
@@ -1230,19 +1235,19 @@ func (scope *Scope) removeForeignKey(field string, dest string) {
 	scope.Raw(fmt.Sprintf(query, scope.QuotedTableName(), scope.quoteIfPossible(keyName))).Exec()
 }
 
-func (scope *Scope) removeIndex(indexName string) {
-	scope.Dialect().RemoveIndex(scope.TableName(), indexName)
+func (scope *Scope) RemoveIndexContext(indexName string) {
```
Comment (ts4z) [inline scope.go:1238]: did we want to make this public?

this has the same signature and replaces the old method. Does it need to be a new method?

**Comment by paulyea** [comment]
both seem to be a pretty simple change to make if we ever decide to push this upstream.

File: main.go (hunk 1 above)
Comment (paulyea) [reply main.go:636]: def a mistake on my part. Will fix

File: main_test.go (hunk 2 above)
Comment (paulyea) [reply main_test.go:198]: hmm...I think this was a find and replace mistake. Will also undo

File: scope.go (hunk 3 above)
Comment (paulyea) [reply scope.go:1238]: Also think this is a find and replace mistake. Will undo

**Pull request merged**